### PR TITLE
More of Object

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/logic/system/partitionmanager.h [hunk 1]
```diff
@@ -158,6 +158,7 @@ class PartitionData : public MemoryPoolObject
 
     void Remove_From_Dirty_Modules(PartitionData **dirtyModules);
     void Make_Dirty(bool b);
+    ObjectShroudStatus Get_Shrouded_Status(int index);
```
Comment (xezon) [inline src/game/logic/system/partitionmanager.h:161]: const?

File: src/game/logic/scriptengine/scriptengine.h [hunk 2]
```diff
@@ -178,6 +178,7 @@ class ScriptEngine : public SubsystemInterface, public SnapShot
 
     void Set_Global_Difficulty(GameDifficulty diff);
     void Set_Use_Object_Difficulty_Bonus(bool bonus) { m_useObjectDifficultyBonuses = bonus; }
+    bool Get_Use_Object_Difficulty_Bonus() { return m_useObjectDifficultyBonuses; }
```
Comment (xezon) [inline src/game/logic/scriptengine/scriptengine.h]: const

File: src/game/logic/object/upgrade/radarupgrade.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    virtual bool Is_Sub_Objects_Upgrade() override;
+    virtual void Upgrade_Implementation() override;
+
+    const RadarUpgradeModuleData *Get_Radar_Upgrade_Module_Data()
+    {
+        return static_cast<const RadarUpgradeModuleData *>(Get_Module_Data());
+    }
+
+    bool Get_Disable_Proof() { return Get_Radar_Upgrade_Module_Data()->m_disableProof; }
```
Comment (xezon) [inline src/game/logic/object/upgrade/radarupgrade.h]: const

File: src/game/logic/object/object.cpp [hunk 4]
```diff
@@ -846,30 +858,47 @@ void Object::Load_Post_Process()
...
-#else
-    return Relationship();
-#endif
+    const Team *team = Get_Team();
+
+    if (team == nullptr || that == nullptr) {
+        return NEUTRAL;
+    }
+
+    if (this->Is_Undetected_Defector()) {
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: `this->` can be omitted.

File: src/game/logic/object/object.cpp [hunk 5]
```diff
@@ -878,29 +907,124 @@ bool Object::Is_In_List(Object **list) const
...
+        Give_Upgrade(upgrade);
+    }
+
+    BodyModuleInterface *body = Get_Body_Module();
+
+    if (body != nullptr) {
+        body->On_Veterancy_Level_Changed(old_level, new_level, play_sound);
+    }
+
+    bool b1 = false;
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: object_hidden

File: src/game/logic/object/object.cpp [hunk 6]
```diff
@@ -878,29 +907,124 @@ bool Object::Is_In_List(Object **list) const
...
+
+    if (!Is_Locally_Controlled() && Get_Status(OBJECT_STATUS_STEALTHED) && !Get_Status(OBJECT_STATUS_DETECTED)
+        && !Get_Status(OBJECT_STATUS_DISGUISED)) {
+        b1 = true;
+    }
+
+    bool b2 = !b1 && new_level > old_level && !Is_KindOf(KINDOF_IGNORED_IN_GUI);
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: show_level_up

File: src/game/logic/object/object.cpp [hunk 7]
```diff
@@ -2021,3 +2346,541 @@ void Object::Notify_Firing_Tracker_Shot_Fired(const Weapon *weapon, ObjectID id)
...
+
+    if (edge && dozer != nullptr && dozer->Get_Current_Task() != DOZER_TASK_INVALID) {
+        dozer->Cancel_Task(dozer->Get_Current_Task());
+    }
+
+    Player *player = Get_Controlling_Player();
+
+    if (player != nullptr) {
+        static NameKeyType radar = g_theNameKeyGenerator->Name_To_Key("RadarUpgrade");
+        RadarUpgrade *upgrade = static_cast<RadarUpgrade *>(Find_Module(radar));
```
Comment (xezon) [inline src/game/logic/object/object.cpp:2419]: const?

File: src/game/logic/object/object.cpp [hunk 8]
```diff
@@ -2021,3 +2346,541 @@ void Object::Notify_Firing_Tracker_Shot_Fired(const Weapon *weapon, ObjectID id)
...
+            }
+        }
+    }
+
+    int energy = Get_Template()->Get_Energy_Production();
+
+    if (energy > 0) {
+        static NameKeyType powerPlant = g_theNameKeyGenerator->Name_To_Key("PowerPlantUpgrade");
+        static NameKeyType overCharge = g_theNameKeyGenerator->Name_To_Key("OverchargeBehavior");
+        PowerPlantUpgrade *power = static_cast<PowerPlantUpgrade *>(Find_Module(powerPlant));
```
Comment (xezon) [inline src/game/logic/object/object.cpp:2437]: const?

File: src/game/logic/object/object.cpp [hunk 9]
```diff
@@ -2021,3 +2346,541 @@ void Object::Notify_Firing_Tracker_Shot_Fired(const Weapon *weapon, ObjectID id)
...
+
+        if (power != nullptr && power->Is_Already_Upgraded()) {
+            energy += Get_Template()->Get_Energy_Bonus();
+        }
+
+        OverchargeBehavior *charge = static_cast<OverchargeBehavior *>(Find_Module(overCharge));
```
Comment (xezon) [inline src/game/logic/object/object.cpp:2443]: const?

File: src/game/logic/object/object.cpp [hunk 10]
```diff
@@ -2021,3 +2346,541 @@ void Object::Notify_Firing_Tracker_Shot_Fired(const Weapon *weapon, ObjectID id)
...
+
+    for (int i = 0; i < DISABLED_TYPE_COUNT; i++) {
+        if (Get_Disabled_State(static_cast<DisabledType>(i)) && frame >= m_disabledStateFrames[i]) {
+            Clear_Disabled(static_cast<DisabledType>(i));
+            m_disabledStates.Set(i, false);
+        }
+    }
+}
+
+void Object::Attempt_Healing(float f, const Object *obj)
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: amount

File: src/game/logic/object/object.cpp [hunk 11]
```diff
@@ -2021,3 +2346,541 @@ void Object::Notify_Firing_Tracker_Shot_Fired(const Weapon *weapon, ObjectID id)
...
+            id = OBJECT_UNK;
+        }
+
+        info.m_in.m_sourceID = id;
+        info.m_in.m_amount = f;
+        body->Attempt_Healing(&info);
+    }
+}
+
+void Object::On_Removed_From(Object *removed)
```
Comment (xezon) [inline src/game/logic/object/object.cpp:2489]: `removed` is unused. Is function body correct?

File: src/game/logic/object/object.cpp [hunk 12]
```diff
@@ -1009,25 +1131,111 @@ bool Object::Is_Mass_Selectable() const
...
+                Handle_Partition_Cell_Maintenance();
+            }
+        }
+
+        if (m_status.Test(OBJECT_STATUS_UNDER_CONSTRUCTION) != bits.Test(OBJECT_STATUS_UNDER_CONSTRUCTION)) {
+            SimpleObjectIterator *iter = g_thePartitionManager->Iterate_Potential_Collisions(
+                Get_Position(), Get_Geometry_Info(), Get_Orientation(), false);
+            MemoryPoolObjectHolder holder(iter);
+
+            for (Object *o = iter->First(); o != nullptr; o = iter->Next()) {
```
Comment (xezon) [inline src/game/logic/object/object.cpp:1224]: obj

File: src/game/logic/system/partitionmanager.h (hunk 1 above)
Comment (jonwil) [reply src/game/logic/system/partitionmanager.h:161]: Nope.

File: src/game/logic/scriptengine/scriptengine.h (hunk 2 above)
Comment (jonwil) [reply src/game/logic/scriptengine/scriptengine.h]: fixed.

File: src/game/logic/object/upgrade/radarupgrade.h (hunk 3 above)
Comment (jonwil) [reply src/game/logic/object/upgrade/radarupgrade.h]: fixed.

File: src/game/logic/object/object.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 6 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:2419]: nope.

File: src/game/logic/object/object.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:2437]: nope.

File: src/game/logic/object/object.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:2443]: nope.

File: src/game/logic/object/object.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:2489]: Yes.

File: src/game/logic/object/object.cpp (hunk 10 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp [hunk 13]
```diff
@@ -1046,25 +1254,154 @@ void Object::Clear_Model_Condition_State(ModelConditionFlagType a)
...
+            if (!Is_Disabled()) {
+                On_Disabled_Edge(false);
+            }
+
+            return true;
+        } else {
+            return false;
+        }
+    } else {
+        captainslog_dbgassert(false, "Invalid disabled type value %d specified -- doesn't not exist!", type);
```
Comment (xezon) [inline src/game/logic/object/object.cpp:1371]:
```
doesn't not
```
double negative

**Review by xezon — COMMENTED** [review]
I tested a bit. One problem so far. Not yet clear if introduced by this change.

Shell map looks as usual.
Veterancy level up works.
Veterancy level up animation works.
Producing veteran units from factories works.
Shroud works.
~~USA Search And Destroy Battle Plan range bonus does not work on units.~~
Unit damage works.
Unit damage radar event works.
Unit disable and reenable with subdual, hacker, emp works.
Pausing and resuming special power timers works.
Infantry crushing works.
Weapon upgrades work.
Out-of-power disable works.
Power overcharge works.
Healing works.
Interactions with neutral buildings work.
Units not controllable when disabled works.
Object changing shroud range works.

**Review by xezon — APPROVED** [review]
Most things this change touches can be tested and those things tested appeared to work correctly. No crashes. Tested in RelWithDebInfo.

**Pull request merged**